Projects/soupstone.md

```diff
@@ -3,19 +3,95 @@
 
 ## Areas of research
 
+### No Debugger
 
-## Note logs
+In Rust, it is very common to use an external debugging probe via SWD, like a [J-Link] or [DAPLink]. This makes it really easy to reprogram the device, and even get logs via something like [probe-run] and [defmt]. These are made possible using [probe-rs]'s
+
+On one hand, this is super powerful. It makes a lot of the initial board bringup painless, and every Cortex-M board supports it.
+
+On the other hand, not every chip has as good of a debugger story. ESP32 has a couple of different debugging link types. RISC-V processors don't typically have a standard debug interface yet.
+
+On top of that, boards like the XIAO don't have the debugging pins on the header, which means you need pogo pins or solder wires onto the bottom of the board. For people without a soldering iron or the skills to solder to a board or design their own board, this might not be an option at all.
+
+In short: I want to play around with different ways of iterating on firmware, without using a debugger. I hope this opens up more opportunities when you don't have a chip, or a board, or the materials to be able to lean on the debugger to make magic happen.
+
+I plan on using the built-in USB support of the nRF52840, but a similar approach should be possible over UART, or whatever other interfaces are available.
+
+[J-Link]: https://www.segger.com/products/debug-probes/j-link/
+[DAPLink]: https://daplink.io/
+[probe-run]: https://github.com/knurling-rs/probe-run
+[defmt]: https://github.com/knurling-rs/defmt/
+[probe-rs]: https://probe.rs/
+
+### RAM Loader
 
+I got really in my head about having "the perfect bootloader". It can be really hard to make a "bullet proof" bootloader, and flexibily support flashing to both internal flash, and external flash.
+
+To break the "[analysis paralysis]", I decided to make the dumbest persistent loader I could.
+
+This meant:
+
+* It could only write to RAM, not flash
+* A very simple, USB-serial link, using [postcard]
+* Nothing I write with the loader will survive a reboot
+
+Most of the time when you are writing early code, or rapidly iterating, you don't really care about the code running when it isn't attached to your computer. If you have enough RAM to fit both the application, and the RAM the application needs, that means you can skip the slowest part of "flashing" a device: erasing and writing the flash memory.
+
+Luckily with the nRF52840, you have 256KiB of RAM, which is usually plenty for both application and running RAM. For reference, I often work with [STM32G0] chips, which only have 16KiB-64KiB of Flash, and 8KiB of RAM, so the nRF has more space than I'd ever see on that entire device!
+
+At the moment, I have a working [stage zero] RAM loader working, which makes it pretty quick to reboot, flash, reboot, and run brand new code.
+
+[analysis paralysis]: https://en.wikipedia.org/wiki/Analysis_paralysis
+[postcard]: ../postcard.md
```

Comment:
just for my understanding of how links work here- this links to the postcard part of this entire project? the .. is to go back a folder in the files and then the name of the file?

Reply:
Oops! Yes, `../postcard.md` means "go up one folder then link to `postcard.md`. But this is wrong, because `postcard.md` is in the same folder as `soupstone.md`, so what I *should* have put here is `./postcard.md`, which means "in the same folder as we are in right now".